lib/interceptor/decompress.js

```diff
@@ -0,0 +1,123 @@
+    super(handler)
+    this.#handler = handler
+    this.#opts = opts
+  }
+
+  onConnect (abort) {
+    this.#inputStream = null
+    this.#trailers = null
+    return this.#handler.onConnect(abort)
+  }
+
+  #onDecompressStreamFinished (err) {
+    if (err) {
+      this.#handler.onError(err)
+    } else {
+      this.#handler.onComplete(this.#trailers)
+    }
+  }
+
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    const parsedHeaders = util.parseHeaders(rawHeaders)
+    const contentEncoding = parsedHeaders['content-encoding']
+    const encodings = contentEncoding ? contentEncoding.split(',').map(e => e.trim().toLowerCase()) : []
+
+    const { method } = this.#opts
+
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
+    if (encodings.length !== 0 && method !== 'HEAD' && method !== 'CONNECT' && !nullBodyStatus.includes(statusCode)) {
+      const decoders = []
+      for (let i = 0; i < encodings.length; ++i) {
+        const encoding = encodings[i]
+        // https://www.rfc-editor.org/rfc/rfc9112.html#section-7.2
+        if (encoding === 'x-gzip' || encoding === 'gzip') {
+          decoders.push(zlib.createGunzip({
+            // Be less strict when decoding compressed responses, since sometimes
+            // servers send slightly invalid responses that are still accepted
+            // by common browsers.
+            // Always using Z_SYNC_FLUSH is what cURL does.
+            flush: zlib.constants.Z_SYNC_FLUSH,
+            finishFlush: zlib.constants.Z_SYNC_FLUSH
+          }))
+        } else if (encoding === 'deflate') {
+          decoders.push(createInflate())
+        } else if (encoding === 'br') {
+          decoders.push(zlib.createBrotliDecompress())
+        } else {
+          decoders.length = 0
+          break
+        }
+      }
+
+      if (decoders.length !== 0) {
+        const [firstDecoder, ...restDecoders] = decoders
+        this.#inputStream = firstDecoder
+
+        if (restDecoders.length !== 0) {
+          pipeline(
+            this.#inputStream,
+            ...restDecoders,
+            this.#onDecompressStreamFinished.bind(this)
+          ).on('data', (chunk) => this.#handler.onData(chunk))
+        } else {
+          finished(
+            this.#inputStream.on('data', (chunk) => this.#handler.onData(chunk)),
+            this.#onDecompressStreamFinished.bind(this)
+          )
+        }
+      }
+    }
+
+    return this.#handler.onHeaders(
+      statusCode,
+      rawHeaders,
+      resume,
+      statusMessage
+    )
+  }
+
+  onData (chunk) {
+    if (this.#inputStream) {
+      return this.#inputStream.write(chunk)
+    }
+    return this.#handler.onData(chunk)
+  }
+
+  onComplete (trailers) {
+    if (this.#inputStream) {
+      this.#trailers = trailers
+      this.#inputStream.end()
+      return
+    }
+
+    return this.#handler.onComplete(trailers)
+  }
+}
+
+function createDecompressionInterceptor () {
```

Comment:
We might need to support add the hinting of supported encoding in case implementers wants to limit this (I can only think for performance reasons).
They can pass a set of encodings supported, and the decompress decorator hints it using [`accept-encoding`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding)